core/app/models/spree/asset.rb

```diff
@@ -1,9 +1,17 @@
 module Spree
-  Paperclip.interpolates :viewable_id do |attachment, _style|
-    attachment.instance.viewable_id
+  if Rails.application.config.use_paperclip
+    ActiveSupport::Deprecation.warn(<<-EOS, caller)
+     Paperclip is deprecated, and will be removed in Spree 4.0.
+     Please migrate to ActiveStorage, to avoid problems after update
```

Comment:
`Please migrate to ActiveStorage, to avoid problems after update`

->

```
Please migrate to ActiveStorage, to avoid problems after update
https://github.com/thoughtbot/paperclip/blob/master/MIGRATING.md
```